core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,217 @@
+  }
+
+  public UpdateSnapshotReferencesOperation removeTag(String name) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef ref = updatedRefs.remove(name);
+    Preconditions.checkArgument(ref != null, "No such tag with name %s", name);
+    Preconditions.checkArgument(ref.isTag(), "Ref with name %s is a branch not a tag", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Branch name cannot be null");
+    Preconditions.checkArgument(!name.equals(SnapshotRef.MAIN_BRANCH), "Cannot rename main branch");
+    SnapshotRef ref = updatedRefs.get(name);
+    Preconditions.checkArgument(ref != null, "No such ref with name %s", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef existing = updatedRefs.put(newName, ref);
+    ValidationException.check(existing == null, "Ref %s already exists", newName);
+    updatedRefs.remove(name, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    Preconditions.checkArgument(ref != null, "Branch %s does not exist", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(ref, snapshotId).build();
+    updatedRefs.put(name, updatedRef);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, String source) {
+    return replaceBranch(name, source, false);
+  }
+
+  public UpdateSnapshotReferencesOperation fastForward(String name, String source) {
+    return replaceBranch(name, source, true);
+  }
+
+  private UpdateSnapshotReferencesOperation replaceBranch(String name, String source, boolean fastForward) {
+    Preconditions.checkNotNull(name, "Target branch cannot be null");
+    Preconditions.checkNotNull(source, "Source ref cannot be null");
+    SnapshotRef sourceRef = updatedRefs.get(source);
+    SnapshotRef refToUpdate = updatedRefs.get(name);
+    Preconditions.checkArgument(sourceRef != null, "Ref %s does not exist", source);
+    Preconditions.checkArgument(refToUpdate != null, "Branch %s does not exist", name);
+    Preconditions.checkArgument(refToUpdate.isBranch(), "Ref with name %s is a tag not a branch", name);
+
+    // Nothing to replace
+    if (sourceRef.snapshotId() == refToUpdate.snapshotId()) {
+      return this;
+    }
+
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(refToUpdate, sourceRef.snapshotId()).build();
+
+    if (fastForward) {
+      Set<Long> sourceAncestors = Sets.newHashSet(SnapshotUtil.ancestorIds(
+          base.snapshot(sourceRef.snapshotId()), base::snapshot));
```

Comment:
Why does this build a new HashSet from the ancestor IDs? Does `contains` not work on what that already returns? (Curious, not a blocker.)

Reply:
SnapshotUtil.ancestorIds returns a List. We convert to the hashset just so the contains later is O(1) instead of O(n). But considering we already did an o(n) operation to take the list element and put it in a set, it probably does not matter. We can also just traverse the ancestors ourselves and not even need a separate data structure for storing it since all we care about is if target is an ancestor of source.

Reply:
Yep, that's what I was thinking. No sense in traversing the list to make a set when we're going to immediately discard it. This is actually slower because we can at least stop when traversing the list!

Reply:
Yup, we can just use SnapshotUtil.isAncestorOf which does this for us